src/impl_owned_array.rs

```diff
@@ -59,3 +69,558 @@ where
+    /// Appending automatically changes memory layout of the array so that it is appended to
+    /// along the "growing axis".
+    ///
+    /// Ensure appending is efficient by for example starting appending an empty array and
+    /// always appending along the same axis. For rows, ndarray's default layout is efficient for
+    /// appending.
+    ///
+    /// Notice that an empty array (where it has an axis of length zero) is the simplest starting
+    /// point. The amortized average complexity of the append is O(m) where *m* is the length of
+    /// the row.
+    ///
+    /// ```rust
+    /// use ndarray::{Array, ArrayView, array};
+    ///
+    /// // create an empty array and append
+    /// let mut a = Array::zeros((0, 4));
+    /// a.append_row(ArrayView::from(&[ 1.,  2.,  3.,  4.])).unwrap();
+    /// a.append_row(ArrayView::from(&[-1., -2., -3., -4.])).unwrap();
+    ///
+    /// assert_eq!(
+    ///     a,
+    ///     array![[ 1.,  2.,  3.,  4.],
+    ///            [-1., -2., -3., -4.]]);
+    /// ```
+    pub fn append_row(&mut self, row: ArrayView<A, Ix1>) -> Result<(), ShapeError>
+    where
+        A: Clone,
+    {
+        self.append(Axis(0), row.insert_axis(Axis(0)))
+    }
+
+    /// Append a column to an array
+    ///
+    /// ***Errors*** with a shape error if the length of the column does not match the length of
+    /// the columns in the array.<br>
+    ///
+    /// The memory layout of the `self` array matters for ensuring that the append is efficient.
+    /// Appending automatically changes memory layout of the array so that it is appended to
+    /// along the "growing axis".
+    ///
+    /// Ensure appending is efficient by for example starting appending an empty array and
+    /// always appending along the same axis. For columns, column major ("F") memory layout is
+    /// efficient for appending.
+    ///
+    /// Notice that an empty array (where it has an axis of length zero) is the simplest starting
+    /// point. The amortized average complexity of the append is O(m) where *m* is the length of
+    /// the row.
+    ///
+    /// ```rust
+    /// use ndarray::{Array, ArrayView, array};
+    ///
+    /// // create an empty array and append
+    /// let mut a = Array::zeros((2, 0));
+    /// a.append_column(ArrayView::from(&[1., 2.])).unwrap();
+    /// a.append_column(ArrayView::from(&[-1., -2.])).unwrap();
+    ///
+    /// assert_eq!(
+    ///     a,
+    ///     array![[1., -1.],
+    ///            [2., -2.]]);
+    /// ```
+    pub fn append_column(&mut self, column: ArrayView<A, Ix1>) -> Result<(), ShapeError>
+    where
+        A: Clone,
+    {
+        self.append(Axis(1), column.insert_axis(Axis(1)))
+    }
+}
+
+impl<A, D> Array<A, D>
+    where D: Dimension
+{
+    /// Move all elements from self into `new_array`, which must be of the same shape but
+    /// can have a different memory layout. The destination is overwritten completely.
+    ///
```

Comment:
We should also mention in the docs that if an element panics on drop, some elements may be leaked. (This can occur in `drop_unreachable_elements_slow`.)

Reply:
Hm right. Maybe that a difference to what drop_in_place on [T] does (I think it attempts to continue dropping?) Good call, I think we can keep doing that, but document it.

Reply:
I'm incorporating that, but it seems so general and expected that maybe it doesn't need mentioning? It is nice to be so overly specific when we can, maybe.